core/vm/contracts.go

```diff
@@ -26,63 +26,44 @@ import (
-	Gas func(l int) *big.Int
-	fn  func(in []byte) []byte
+// Precompiled contract is the basic interface for native Go contracts. The implementation
+// requires a deterministic gas count based on the input size of the Run method of the
+// contract.
+type PrecompiledContract interface {
+	RequiredGas(inputSize int) *big.Int // RequiredPrice calculates the contract gas use
+	Run(input []byte) []byte            // Run runs the precompiled contract
 }
 
-// Call calls the native function
-func (self PrecompiledAccount) Call(in []byte) []byte {
-	return self.fn(in)
+// Precompiled contains the default set of ethereum contracts
+var PrecompiledContracts = map[common.Address]PrecompiledContract{
+	common.BytesToAddress([]byte{1}): &ecrecover{},
+	common.BytesToAddress([]byte{2}): &sha256{},
+	common.BytesToAddress([]byte{3}): &ripemd160{},
+	common.BytesToAddress([]byte{4}): &dataCopy{},
 }
 
-// Precompiled contains the default set of ethereum contracts
-var Precompiled = PrecompiledContracts()
-
-// PrecompiledContracts returns the default set of precompiled ethereum
-// contracts defined by the ethereum yellow paper.
-func PrecompiledContracts() map[string]*PrecompiledAccount {
-	return map[string]*PrecompiledAccount{
-		// ECRECOVER
-		string(common.LeftPadBytes([]byte{1}, 20)): &PrecompiledAccount{func(l int) *big.Int {
-			return params.EcrecoverGas
-		}, ecrecoverFunc},
-
-		// SHA256
-		string(common.LeftPadBytes([]byte{2}, 20)): &PrecompiledAccount{func(l int) *big.Int {
-			n := big.NewInt(int64(l+31) / 32)
-			n.Mul(n, params.Sha256WordGas)
-			return n.Add(n, params.Sha256Gas)
-		}, sha256Func},
-
-		// RIPEMD160
-		string(common.LeftPadBytes([]byte{3}, 20)): &PrecompiledAccount{func(l int) *big.Int {
-			n := big.NewInt(int64(l+31) / 32)
-			n.Mul(n, params.Ripemd160WordGas)
-			return n.Add(n, params.Ripemd160Gas)
-		}, ripemd160Func},
-
-		string(common.LeftPadBytes([]byte{4}, 20)): &PrecompiledAccount{func(l int) *big.Int {
-			n := big.NewInt(int64(l+31) / 32)
-			n.Mul(n, params.IdentityWordGas)
-
-			return n.Add(n, params.IdentityGas)
-		}, memCpy},
+// RunPrecompile runs and evaluate the output of a precompiled contract defined in contracts.go
+func RunPrecompiled(p PrecompiledContract, input []byte, contract *Contract) (ret []byte, err error) {
+	gas := p.RequiredGas(len(input))
+	if contract.UseGas(gas.Uint64()) {
+		ret = p.Run(input)
+
+		return ret, nil
+	} else {
+		return nil, OutOfGasError
 	}
 }
 
-func sha256Func(in []byte) []byte {
-	return crypto.Sha256(in)
-}
+// ECRECOVER implemented as a native contract
+type ecrecover struct{}
 
-func ripemd160Func(in []byte) []byte {
-	return common.LeftPadBytes(crypto.Ripemd160(in), 32)
+func (c *ecrecover) RequiredGas(inputSize int) *big.Int {
+	return params.EcrecoverGas
 }
 
-const ecRecoverInputLength = 128
+func (c *ecrecover) Run(in []byte) []byte {
+	const ecRecoverInputLength = 128
 
-func ecrecoverFunc(in []byte) []byte {
 	in = common.RightPadBytes(in, 128)
```

Comment:
I actually only moved the `const`. But you are right, it should use the constant.